{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy\n",
    "import datetime\n",
    "import re\n",
    "import collections\n",
    "from collections import OrderedDict\n",
    "import math\n",
    "import allel\n",
    "import pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# new_file_str = 'binom_call-merged_normal.vcf'\n",
    "new_vcf_str = 'temp.vcf' \n",
    "# new_vcf_str = 'variant_call_resaults/'\n",
    "ref_vcf_str = '/sbgenomics/project-files/final_project_files/final_merged_pileup.called.vcf'\n",
    "# ref_vcf_str = 'test_ref.vcf'\n",
    "output_file_temp_str = 'temp_metrics_files/temp_metrics'\n",
    "\n",
    "temp_new_vcf_str = 'temp_metrics_files/temp_new.vcf'\n",
    "temp_ref_vcf_str = 'temp_metrics_files/temp_ref.vcf'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {},
   "outputs": [],
   "source": [
    "def vcf_to_df(vcf_file):\n",
    "    callset = allel.read_vcf(vcf_file, fields=['variants/CHROM', 'variants/POS', 'variants/REF', 'variants/ALT',  'calldata/GT'])\n",
    "    \n",
    "    df_chrom = pandas.DataFrame.from_dict(callset['variants/CHROM'])#, callset['variants/POS'], callset['variants/REF'])\n",
    "    df_chrom = df_chrom.rename(columns={0:'CHROM' }) #, 1:'POS', 2:'REF'\n",
    "    df_pos = pandas.DataFrame.from_dict(callset['variants/POS'], dtype='int64')\n",
    "    df_pos = df_pos.rename(columns={0:'POS' })\n",
    "    df_ref = pandas.DataFrame.from_dict(callset['variants/REF'])\n",
    "    df_ref = df_ref.rename(columns={0:'REF'})\n",
    "    df_alts = pandas.DataFrame.from_dict(callset['variants/ALT'])\n",
    "    df_alts = df_alts.rename(columns={0:'ALT_0', 1:'ALT_1'})\n",
    "    df_gt = pandas.DataFrame.from_dict(callset['calldata/GT'][:,0])\n",
    "    df_gt = df_gt.rename(columns={0:'GT_0', 1:'GT_1'})\n",
    "    frames = [ df_chrom, df_pos, df_ref, df_alts, df_gt]\n",
    "    vcf = pandas.concat(frames, axis=1)\n",
    "    return vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def declare_variants_only_with_changes(gt0, gt1, ref, alt0, alt1, pos):\n",
    "#     variants = []\n",
    "#     for i in range(0,len(gt0)):\n",
    "# #         print(gt0[i])\n",
    "# #         print(gt1[i])\n",
    "# #         print(ref[i])\n",
    "# #         print(alt0[i])\n",
    "# #         print(alt1[i])\n",
    "# #         if(gt0[i] == 0 and gt1[i] == 0):\n",
    "# #             variants.append(ref[i])\n",
    "#         if(gt0[i] == 0 and gt1[i] == 1):\n",
    "#             s = str(pos[i]) + '-' + str(ref[i]) + str(alt0[i])\n",
    "#             variants.append(s)\n",
    "# #             variants.append(''.join([ref[i],alt0[i]]))\n",
    "#         if(gt0[i] == 1 and gt1[i] == 1):\n",
    "#             s = str(pos[i]) + '-' + alt0[i]\n",
    "#             variants.append(s)\n",
    "# #             variants.append(alt0[i])\n",
    "#         if(gt0[i] == 1 and gt1[i] == 2):\n",
    "#             s = str(pos[i]) + '-' + str(alt0[i]) + str(alt1[i])\n",
    "#             variants.append(s)\n",
    "# #             variants.append(''.join([alt0[i],alt1[i]]))\n",
    "            \n",
    "#     return variants\n",
    "\n",
    "# def declare_variants_without_changes(gt0, gt1, ref, alt0, alt1, pos):\n",
    "#     variants = []\n",
    "#     for i in range(0,len(gt0)):\n",
    "# #         print(gt0[i])\n",
    "# #         print(gt1[i])\n",
    "# #         print(ref[i])\n",
    "# #         print(alt0[i])\n",
    "# #         print(alt1[i])\n",
    "#         if(gt0[i] == 0 and gt1[i] == 0):\n",
    "#             s = str(pos[i]) + '-' + ref[i]\n",
    "#             variants.append(s)\n",
    "# #             variants.append(ref[i])\n",
    "# #         if(gt0[i] == 0 and gt1[i] == 1):\n",
    "# #             variants.append(''.join([ref[i],alt0[i]]))\n",
    "# #         if(gt0[i] == 1 and gt1[i] == 1):\n",
    "# #             variants.append(alt0[i])\n",
    "# #         if(gt0[i] == 1 and gt1[i] == 2):\n",
    "# #             variants.append(''.join([alt0[i],alt1[i]]))\n",
    "            \n",
    "#     return variants\n",
    "        \n",
    "def declare_variants_only_with_changes(gt0, gt1, ref, alt0, alt1, pos):\n",
    "    variants = []\n",
    "    for i in range(0,len(gt0)):\n",
    "    #         print(gt0[i])\n",
    "    #         print(gt1[i])\n",
    "    #         print(ref[i])\n",
    "    #         print(alt0[i])\n",
    "    #         print(alt1[i])\n",
    "    #         if(gt0[i] == 0 and gt1[i] == 0):\n",
    "    #             variants.append(ref[i])\n",
    "        if(gt0[i] == 0 and gt1[i] == 1):\n",
    "            s = str(pos) + '-' + str(ref[i]) + str(alt0[i])\n",
    "            variants.append(s)\n",
    "    #             variants.append(''.join([ref[i],alt0[i]]))\n",
    "        if(gt0[i] == 1 and gt1[i] == 1):\n",
    "            s = str(pos) + '-' + alt0[i]\n",
    "            variants.append(s)\n",
    "    #             variants.append(alt0[i])\n",
    "        if(gt0[i] == 1 and gt1[i] == 2):\n",
    "            s = str(pos) + '-' + str(alt0[i]) + str(alt1[i])\n",
    "            variants.append(s)\n",
    "    #             variants.append(''.join([alt0[i],alt1[i]]))\n",
    "\n",
    "    return variants\n",
    "\n",
    "def declare_variants_without_changes(gt0, gt1, ref, alt0, alt1, pos):\n",
    "    variants = []\n",
    "    for i in range(0,len(gt0)):\n",
    "    #         print(gt0[i])\n",
    "    #         print(gt1[i])\n",
    "    #         print(ref[i])\n",
    "    #         print(alt0[i])\n",
    "    #         print(alt1[i])\n",
    "        if(gt0[i] == 0 and gt1[i] == 0):\n",
    "            s = str(pos) + '-' + ref[i]\n",
    "            variants.append(s)\n",
    "    #             variants.append(ref[i])\n",
    "    #         if(gt0[i] == 0 and gt1[i] == 1):\n",
    "    #             variants.append(''.join([ref[i],alt0[i]]))\n",
    "    #         if(gt0[i] == 1 and gt1[i] == 1):\n",
    "    #             variants.append(alt0[i])\n",
    "    #         if(gt0[i] == 1 and gt1[i] == 2):\n",
    "    #             variants.append(''.join([alt0[i],alt1[i]]))\n",
    "\n",
    "    return variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_basic_metrics(new_dataset, ref_dataset):\n",
    "    new_position_match = new_dataset['POS'].isin(ref_dataset['POS']).value_counts()\n",
    "    ref_position_match = ref_dataset['POS'].isin(new_dataset['POS']).value_counts()\n",
    "    \n",
    "#     TP = new_position_match[True]\n",
    "#     FP = new_position_match[False]\n",
    "#     FN = ref_position_match[False]\n",
    "    TP = 0\n",
    "    FP = 0\n",
    "    FN = 0\n",
    "    \n",
    "    if True in new_position_match.index: TP = new_position_match[True]\n",
    "    if False in new_position_match.index: FP = new_position_match[False]\n",
    "    if False in ref_position_match.index: FN = ref_position_match[False]\n",
    "    \n",
    "    precision = TP/(TP+FP)\n",
    "    recall = TP/(TP+FN)\n",
    "    f1_score = 2*precision*recall/(precision+recall)\n",
    "         \n",
    "    return TP, FP, FN, precision, recall, f1_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_advance_metrics(new_dataset, ref_dataset):\n",
    "    new_position_df = new_dataset['POS'].isin(ref_dataset['POS'])\n",
    "    ref_position_df = ref_dataset['POS'].isin(new_dataset['POS'])\n",
    "    \n",
    "    new_position_cnt = new_dataset['POS'].isin(ref_dataset['POS']).value_counts()\n",
    "    ref_position_cnt = ref_dataset['POS'].isin(new_dataset['POS']).value_counts()\n",
    "    \n",
    "    TP = 0\n",
    "    FP = 0\n",
    "    FN = 0\n",
    "    precision = 0\n",
    "    recall = 0\n",
    "    f1_score = 0\n",
    "    \n",
    "    if False in ref_position_cnt.index: FN += ref_position_cnt[False]\n",
    "    if False in new_position_cnt.index: FP += new_position_cnt[False]\n",
    "    \n",
    "#     print (FP)\n",
    "#     print (FN)\n",
    "    \n",
    "    #indexes of samples from ref file\n",
    "    ref_match_new_pos = ref_position_df.index[ref_position_df == True].tolist()\n",
    "#     print (ref_match_new_pos)\n",
    "    \n",
    "     #indexes of samples from new file\n",
    "    new_match_rf_pos = new_position_df.index[new_position_df == True].tolist()\n",
    "#     print (new_match_rf_pos)\n",
    "    \n",
    "    for i in range(0, len(new_match_rf_pos)):\n",
    "\n",
    "        \n",
    "        new_gt0 = new_dataset['GT_0'].iloc[new_match_rf_pos].values\n",
    "        new_gt1 = new_dataset['GT_1'].iloc[new_match_rf_pos].values\n",
    "        new_ref = new_dataset['REF'].iloc[new_match_rf_pos].values\n",
    "        new_alt0 = new_dataset['ALT_0'].iloc[new_match_rf_pos].values\n",
    "        new_alt1 = new_dataset['ALT_1'].iloc[new_match_rf_pos].values\n",
    "    #     new_pos = new_dataset['POS'].iloc[new_match_rf_pos].values\n",
    "\n",
    "        new_variants = pandas.Series(declare_variants_only_with_changes(new_gt0, new_gt1, new_ref, new_alt0, new_alt1, 0))\n",
    "        new_non_variants = pandas.Series(declare_variants_without_changes(new_gt0, new_gt1, new_ref, new_alt0, new_alt1, 0))\n",
    "        print('new_variants')\n",
    "        print(new_variants)\n",
    "        print('new_non_variants')\n",
    "        print(new_non_variants)\n",
    "\n",
    "    ref_gt0 = ref_dataset['GT_0'].iloc[ref_match_new_pos].values\n",
    "    ref_gt1 = ref_dataset['GT_1'].iloc[ref_match_new_pos].values\n",
    "    ref_ref = ref_dataset['REF'].iloc[ref_match_new_pos].values\n",
    "    ref_alt0 = ref_dataset['ALT_0'].iloc[ref_match_new_pos].values\n",
    "    ref_alt1 = ref_dataset['ALT_1'].iloc[ref_match_new_pos].values\n",
    "#     ref_pos = new_dataset['POS'].iloc[ref_match_new_pos].values\n",
    "    \n",
    "    ref_variants = pandas.Series(declare_variants_only_with_changes(ref_gt0, ref_gt1, ref_ref, ref_alt0, ref_alt1, 0))\n",
    "    ref_non_variants = pandas.Series(declare_variants_without_changes(ref_gt0, ref_gt1, ref_ref, ref_alt0, ref_alt1, 0))\n",
    "    print('ref_variants')\n",
    "    print(ref_variants)\n",
    "    print('ref_non_variants')\n",
    "    print(ref_non_variants)\n",
    "\n",
    "    new_call_match_var = new_variants.isin(ref_variants).value_counts()\n",
    "    new_call_match_non_var = new_non_variants.isin(ref_non_variants).value_counts()\n",
    "    ref_call_match_var = ref_variants.isin(new_variants).value_counts()\n",
    "    ref_call_match_non_var = ref_non_variants.isin(new_non_variants).value_counts()\n",
    "\n",
    "    if True in new_call_match_var.index: TP += new_call_match_var[True]\n",
    "    if True in new_call_match_non_var.index: TP += new_call_match_non_var[True]\n",
    "    if False in new_call_match_var.index: FP += new_call_match_var[False]\n",
    "    if False in new_call_match_non_var.index: FP += new_call_match_non_var[False]\n",
    "    if False in ref_call_match_var.index: FN += ref_call_match_var[False]\n",
    "    if False in ref_call_match_non_var.index: FN += ref_call_match_non_var[False]\n",
    "        \n",
    "    print (TP)\n",
    "    print (FP)\n",
    "    print (FN)\n",
    "    \n",
    "    precision = TP/(TP+FP)\n",
    "    recall = TP/(TP+FN)\n",
    "    f1_score = 2*precision*recall/(precision+recall)\n",
    "    \n",
    "    \n",
    "    return TP, FP, FN, precision, recall, f1_score\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(new_file, ref_file):\n",
    "\n",
    "    new_callset = vcf_to_df(new_file)\n",
    "    ref_callset = vcf_to_df(ref_file)\n",
    "    \n",
    "#     print(new_callset)\n",
    "#     print(ref_callset)\n",
    "    \n",
    "#     print(calc_basic_metrics(new_callset, ref_callset))\n",
    "#     print(calc_advance_metrics(new_callset, ref_callset))\n",
    "    \n",
    "    return calc_advance_metrics(new_callset, ref_callset)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 13, 15, 16, 17, 18, 19]\n",
      "new_variants\n",
      "0     0-A\n",
      "1    0-AG\n",
      "dtype: object\n",
      "new_non_variants\n",
      "0     0-C\n",
      "1     0-T\n",
      "2     0-T\n",
      "3     0-T\n",
      "4     0-C\n",
      "5     0-A\n",
      "6     0-A\n",
      "7     0-A\n",
      "8     0-G\n",
      "9     0-A\n",
      "10    0-T\n",
      "11    0-C\n",
      "12    0-T\n",
      "13    0-A\n",
      "dtype: object\n",
      "ref_variants\n",
      "Series([], dtype: float64)\n",
      "ref_non_variants\n",
      "0     0-C\n",
      "1     0-T\n",
      "2     0-T\n",
      "3     0-T\n",
      "4     0-T\n",
      "5     0-C\n",
      "6     0-A\n",
      "7     0-A\n",
      "8     0-A\n",
      "9     0-A\n",
      "10    0-G\n",
      "11    0-A\n",
      "12    0-T\n",
      "13    0-C\n",
      "14    0-T\n",
      "15    0-A\n",
      "dtype: object\n",
      "14\n",
      "11\n",
      "5\n",
      "Processing done for variant_call_results/binom_p0.5_variant_call.vcf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.6/site-packages/allel/io/vcf_read.py:239: UserWarning: empty CHROM; field: EOL; variant: 23 (:0)\n",
      "  chunks = [d[0] for d in it]\n",
      "/opt/conda/lib/python3.6/site-packages/allel/io/vcf_read.py:1516: UserWarning: invalid INFO header: '##INFO=<ID=VDB,Number=1,Type=Float,Description=\"Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)\",Version=\"3\">\\n'\n",
      "  warnings.warn('invalid INFO header: %r' % header)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "sheet = pandas.DataFrame(columns=['VCF name', 'Binomial probability', 'TP', 'FP', 'FN', 'Precision', 'Recall', 'F1-Score'])\n",
    "# p = [0.5, 0.6, 0.7, 0.8, 0.9, 0.95]\n",
    "p = [0.5]\n",
    "row = 0\n",
    "for i in p:\n",
    "    file = 'variant_call_results/binom_p' + str(i) + '_variant_call.vcf'\n",
    "    test_file = 'test_new.vcf'\n",
    "    test_ref = 'test_ref.vcf'\n",
    "#     TP, FP, FN, prec, rec, f1_s = main(file, ref_vcf_str)\n",
    "    TP, FP, FN, prec, rec, f1_s = main(test_file, test_ref)\n",
    "    sheet.loc[row] = [file] + [i, TP, FP, FN, prec, rec, f1_s]\n",
    "    row += 1\n",
    "    print('Processing done for ' + file)\n",
    "sheet.to_csv('metrics.csv', index =False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VCF name</th>\n",
       "      <th>Binomial probability</th>\n",
       "      <th>TP</th>\n",
       "      <th>FP</th>\n",
       "      <th>FN</th>\n",
       "      <th>Precision</th>\n",
       "      <th>Recall</th>\n",
       "      <th>F1-Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>variant_call_results/binom_p0.5_variant_call.vcf</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2871413</td>\n",
       "      <td>1402</td>\n",
       "      <td>7287</td>\n",
       "      <td>0.999512</td>\n",
       "      <td>0.997469</td>\n",
       "      <td>0.998489</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           VCF name  Binomial probability  \\\n",
       "0  variant_call_results/binom_p0.5_variant_call.vcf                   0.5   \n",
       "\n",
       "        TP    FP    FN  Precision    Recall  F1-Score  \n",
       "0  2871413  1402  7287   0.999512  0.997469  0.998489  "
      ]
     },
     "execution_count": 348,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sheet\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(80, 8)"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_len = ref_callset.shape\n",
    "new_len = new_callset.shape[0]\n",
    "ref_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.6/site-packages/allel/io/vcf_read.py:1516: UserWarning: invalid INFO header: '##INFO=<ID=VDB,Number=1,Type=Float,Description=\"Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)\",Version=\"3\">\\n'\n",
      "  warnings.warn('invalid INFO header: %r' % header)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "with open (new_vcf_str, 'r') as vcf_file, open (ref_vcf_str, 'r') as ref_file, open (temp_new_vcf_str, 'w') as temp_new_vcf, open (temp_ref_vcf_str, 'w') as temp_ref_vcf:\n",
    "    \n",
    "    new_callset = allel.read_vcf('test_new.vcf', fields=['variants/CHROM', 'variants/POS', 'variants/REF', 'variants/ALT',  'calldata/GT'])\n",
    "    ref_callset = allel.read_vcf('test_ref.vcf',  fields=['variants/CHROM', 'variants/POS', 'variants/REF', 'variants/ALT',  'calldata/GT'])\n",
    "#     print(sorted(callset.keys()))\n",
    "\n",
    "# with open (temp_new_vcf_str, 'r') as vcf_file, open (temp_ref_vcf_str, 'r') as ref_file, open (output_file_temp_str, 'w') as temp:\n",
    "#     #TODO\n",
    "#     # go through both files, collect calls from both positions and compare them\n",
    "#     #\n",
    "#     # calculate TP, FP, FN\n",
    "#     # calculate precision, recall and F-score for those numbers\n",
    "#     # \n",
    "    \n",
    "    \n",
    "#     line_num = 0;\n",
    "#     for line in iter(vcf_file.readline, ''):\n",
    "#         line_num+=1 \n",
    "#         if(line[0] != '#'):\n",
    "#             temp_new_vcf.write(line)\n",
    "#     for line in iter(ref_file.readline, ''):\n",
    "#         line_num+=1 \n",
    "#         if(line[0] != '#'):\n",
    "#             temp_ref_vcf.write(line)\n",
    "            \n",
    "#     print('Pre-processing done.')\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
