{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy\n",
    "import datetime\n",
    "import re\n",
    "import collections\n",
    "from collections import OrderedDict\n",
    "import math\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "##\n",
    "## STRING DEFINITIONS - constant strings of files that were used and additional info that couldn't be pulled from the files\n",
    "##\n",
    "\n",
    "reference_file_str = '/sbgenomics/project-files/final_project_files/human_g1k_v37_decoy.fasta'\n",
    "reference_fai_file_str = '/sbgenomics/project-files/final_project_files/human_g1k_v37_decoy.fasta.fai'\n",
    "pileup_file_str = '/sbgenomics/project-files/final_project_files/final_merged_normal_pileup'\n",
    "\n",
    "ref_assembly_str = ',assembly=b37'\n",
    "ref_md5_str = ',md5=bb77e60e9a492fd0172e2b11e6c16afd'\n",
    "ref_species_str = ',species=\\\"Homo sapiens\\\"'\n",
    "ref_taxonomy_str = ',taxonomy=x' #couldn't find what this acctually means\n",
    "\n",
    "\n",
    "# pileup_file_str = 'algo_test_pileup' #rename it to merged_normal-example_human_ref-pileup\n",
    "output_file_temp_str = 'temp.vcf'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "##\n",
    "## VCF CONTIG INFO FUNC - Pulls out information on all contigs from reference .fasta.fai file\n",
    "##\n",
    "\n",
    "def write_contig_info(file):\n",
    "    contig_info =''\n",
    "    for line in file:\n",
    "        field = line.split('\\t')\n",
    "        ID_str = field[0]\n",
    "        len_str = field[1]\n",
    "        contig_info += '##contig=<ID=' + ID_str + \",length=\" + len_str +'>\\n' # + ref_assembly_str + ref_md5_str + ref_species_str + ref_taxonomy_str \n",
    "    return contig_info\n",
    "        \n",
    "#tests\n",
    "with open ('test_index.fai', 'r') as test_file:\n",
    "    assert(write_contig_info(test_file) ==  '##contig=<ID=1,length=249250621>\\n'), 'Incorrect output. Possible incompatible input.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "##\n",
    "## VCF HEADER FUNCTION - Writes out header of the .vcf file\n",
    "##\n",
    "\n",
    "def write_vcf_header(file, fai):\n",
    "    date_time = datetime.datetime.now()\n",
    "    \n",
    "    file_format_str = '##fileformat=VCF4.2\\n'\n",
    "    date_str = '##fileDate=' + date_time.strftime(\"%Y%m%d\") + '\\n'\n",
    "    source_str = '##source=variant_call_binomV0.1\\n'\n",
    "    ref_file_str = '##reference=file://' + reference_file_str + '\\n'\n",
    "    contig_info_str = write_contig_info(fai)    \n",
    "    #filter_string = '##FILTER=<ID=PASS,Description=\\\"\\\"\\n' #TODO check if this is necessary\n",
    "    alt_str = '##ALT=<ID=*,Description=\"Represents allele(s) other than observed.\">\\n'\n",
    "    indel_str = '##INFO=<ID=INDEL,Number=0,Type=Flag,Description=\"Indicates that the variant is an INDEL.\">\\n'\n",
    "    dp_str = '##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Raw read depth\">\\n'\n",
    "    gt_str = '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\\n'\n",
    "    vaf_str = '##FORMAT=<ID=VAF,Number=1,Type=Float,Description=\"Variant frequency in sample\">\\n'\n",
    "    \n",
    "    table_header_str = '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tHCC1143BL\\n'\n",
    "    \n",
    "    file.write(file_format_str + date_str + source_str + ref_file_str + contig_info_str + alt_str + indel_str + dp_str + gt_str + vaf_str + table_header_str)\n",
    "    \n",
    "    fai.close()\n",
    "    return file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "##\n",
    "## CALCULATE VARIANTS CALLS\n",
    "##\n",
    "\n",
    "class Variant:\n",
    "    def __init__(self, symbol, num_occ, num_reads, ref_symbol):\n",
    "        self.var_type, self.var_symbol, self.var_len = self.variant_def(symbol, ref_symbol)\n",
    "        \n",
    "        self.num_occ = num_occ\n",
    "        self.num_reads = num_reads\n",
    "        self.occ_percent = round(float(num_occ/num_reads), 1)\n",
    "        self.VAF = str(self.occ_percent)\n",
    "        self.DP = str(self.num_reads)\n",
    "        \n",
    "    def variant_def(self, symbol, ref_symbol):\n",
    "        if ( len(symbol) == 1):\n",
    "                var_symbol = symbol                \n",
    "                if(symbol == '.'):\n",
    "                    var_type = 'REF'\n",
    "                else:\n",
    "                    var_type = 'SNV'\n",
    "                var_len = 1\n",
    "        else:\n",
    "            if(symbol[0]== '+'):\n",
    "                var_type = 'INS'\n",
    "            else:\n",
    "                var_type = 'DEL'\n",
    "            var_symbol = ref_symbol + ''.join(filter(str.isalpha, symbol))\n",
    "            var_len = int(''.join(filter(str.isdigit, symbol)))\n",
    "            \n",
    "    \n",
    "        return var_type, var_symbol, var_len\n",
    "    \n",
    "# tests\n",
    "test_var = Variant('C', 5, 7, 'A')\n",
    "assert(test_var.var_type == 'SNV'), 'incorrect test_type'\n",
    "assert(test_var.var_symbol  == 'C'), 'incorrect var_symbol'\n",
    "assert(test_var.var_len  == 1), 'incorrect var_len'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.14446"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def binomial(n, k, p):\n",
    "        return (math.factorial(n)/(math.factorial(k)*math.factorial(n-k)))*(p**k)*(1-p)**(n-k)\n",
    "    \n",
    "# test\n",
    "assert(round(binomial(15, 12, 0.5), 5) == round(0.01388549805, 5)),'Incorrect value'\n",
    "round(binomial(15, 15, 0.99), 5)\n",
    "round(binomial(30, 15, 0.5), 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_binomial(var1, var2, p):\n",
    "        res_vars = {}\n",
    "        \n",
    "        res1 = binomial(float(var1.num_reads), float(var1.num_occ), float(p))\n",
    "        res2 = binomial(float(var2.num_reads), float(var2.num_occ), float(p))\n",
    "        res3 = binomial(float(var1.num_reads), float(var1.num_reads), float(p))\n",
    "         \n",
    "        if(res1 >= res2):\n",
    "            if(res1 > res3):\n",
    "                return var1, None\n",
    "            else:\n",
    "                return var1, var2\n",
    "        else: #res1 < res2\n",
    "            if(res2 > res3):\n",
    "                return var2, None\n",
    "            else:\n",
    "                return var1, var2\n",
    "            \n",
    "            \n",
    "#tests\n",
    "test_var1 = Variant('A', 12, 30, 'G')\n",
    "test_var2 = Variant('T', 18, 30, 'G')\n",
    "test_bin_var1, test_bin_var2 = calculate_binomial(test_var1, test_var2, 0.85)\n",
    "assert (test_var1 == test_bin_var1 and test_var2 == test_bin_var2), 'Incorrect binomial distribution variant results'\n",
    "\n",
    "test_var1 = Variant('A', 2, 30, 'G')\n",
    "test_var2 = Variant('T', 28, 30, 'G')\n",
    "test_bin_var1, test_bin_var2 = calculate_binomial(test_var1, test_var2, 0.85)\n",
    "assert (test_var2 == test_bin_var1 and None == test_bin_var2), 'Incorrect binomial distribution variant results'\n",
    "\n",
    "test_var1 = Variant('A', 28, 30, 'G')\n",
    "test_var2 = Variant('T', 2, 30, 'G')\n",
    "test_bin_var1, test_bin_var2 = calculate_binomial(test_var1, test_var2, 0.85)\n",
    "assert (test_var1 == test_bin_var1 and None == test_bin_var2), 'Incorrect binomial distribution variant results'\n",
    "\n",
    "test_var1 = Variant('A', 15, 30, 'G')\n",
    "test_var2 = Variant('T', 15, 30, 'G')\n",
    "test_bin_var1, test_bin_var2 = calculate_binomial(test_var1, test_var2, 0.5)\n",
    "assert (test_var1 == test_bin_var1 and None == test_bin_var2), 'Incorrect binomial distribution variant results'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_read_str(read_str): #raname to calculate_variants\n",
    "        read_str = re.sub('\\^.', '', read_str) #remove caret and next symbol\n",
    "        read_str = re.sub('\\$', '', read_str ) #remove dollar\n",
    "        read_str = re.sub('\\,', '.', read_str) #substitute comma with dot\n",
    "        read_str = read_str.upper() #switch all letters to uppercase\n",
    "        return read_str\n",
    "    \n",
    "\n",
    "assert(prepare_read_str('^<.') == '.'), 'Incorect result'\n",
    "assert(prepare_read_str(',$,$,,At,...,.,,,.aa.,g.,,..,.a.,.^(.,,,a..,') == '....AT..........AA..G.......A.......A...'), 'Incorect result.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "##\n",
    "## STORE SINGLE PILEUP CALLS\n",
    "##\n",
    "\n",
    "class Pileup_line:\n",
    "    def __init__(self, new_line):\n",
    "        self.new_line = new_line\n",
    "        field = new_line.split(\"\\t\")\n",
    "        self.seq_str = field[0] # sequence name string\n",
    "        self.pos_str = field[1] # position string\n",
    "        self.ref_str = field[2] # reference base string\n",
    "        self.cnt_str = field[3] # read count string\n",
    "        self.res_str = field[4] # result string\n",
    "        self.qual_str = field[5] # base quality string\n",
    "        \n",
    "        self.pos = int(self.pos_str)\n",
    "        self.cnt = int(self.cnt_str)\n",
    "        \n",
    "        self.read_vars = collections.OrderedDict()\n",
    "        self.read_vars = {\n",
    "            '.':0,\n",
    "            'A':0,\n",
    "            'T':0,\n",
    "            'C':0,\n",
    "            'G':0\n",
    "        }\n",
    "        \n",
    "        self.var1 = None\n",
    "        self.var2 = None\n",
    "        \n",
    "    def process_read_line(self):\n",
    "        self.find_variants(self.res_str)\n",
    "        return self.call_variants_binomial(self.var1, self.var2)\n",
    "    \n",
    "    \n",
    "    def find_variants (self, read_str):\n",
    "        has_ins = read_str.find('+')\n",
    "        has_del = read_str.find('-')\n",
    "        \n",
    "        read_str = prepare_read_str(read_str)\n",
    "        \n",
    "        if ((has_ins == -1) and (has_del == -1)):            \n",
    "            sym_cnt = collections.Counter(read_str.strip('\"')).most_common()\n",
    "#             print(sym_cnt)\n",
    "            if (len(sym_cnt) == 1):\n",
    "                self.var1 = Variant(sym_cnt[0][0], sym_cnt[0][1], self.cnt, self.ref_str)\n",
    "                self.var2 = None\n",
    "            else:\n",
    "                self.var1 = Variant(sym_cnt[0][0], sym_cnt[0][1], self.cnt, self.ref_str)\n",
    "                self.var2 = Variant(sym_cnt[1][0], sym_cnt[1][1], self.cnt, self.ref_str)\n",
    "        \n",
    "        elif (has_ins != -1 or has_del != -1):\n",
    "#             print(\"INDEL\")\n",
    "            self.prepare_read_vars(read_str)\n",
    "            if(self.read_vars_list[1][1] == 0):\n",
    "                self.var1 = Variant(self.read_vars_list[0][0], self.read_vars_list[0][1], self.cnt, self.ref_str)\n",
    "                self.var2 = None\n",
    "            else:\n",
    "                self.var1 = Variant(self.read_vars_list[0][0], self.read_vars_list[0][1], self.cnt, self.ref_str)\n",
    "                self.var2 = Variant(self.read_vars_list[1][0], self.read_vars_list[1][1], self.cnt, self.ref_str)\n",
    "                \n",
    "#             print(self.var1.var_symbol)\n",
    "#             print(self.var2.var_symbol)\n",
    "                \n",
    "                \n",
    "        \n",
    "    def prepare_read_vars(self, read_str):    \n",
    "        skip_index = 0\n",
    "        \n",
    "        for i in range(0, len(read_str)):\n",
    "            if(i<skip_index) : continue #skip indel symbols\n",
    "             \n",
    "            if(read_str[i] == '+' or read_str[i] == '-'):\n",
    "                num_len = 0\n",
    "                for j in range (i+1, len(read_str)):\n",
    "                    if(read_str[j].isnumeric()): num_len +=1\n",
    "                    else: break\n",
    "                indel_len = int(read_str[i+1 : i+1+num_len])\n",
    "#                 if(read_str[i-1] == '.'): \n",
    "                new_indel = read_str[(i):(i+num_len+indel_len+1)]\n",
    "#                 else:\n",
    "#                     new_ins = read_str[i-1] + read_str[(i+2):(i+2+ins_len)]\n",
    "                #cur_cnt = self.ins_seqs.get(new_ins, 0)\n",
    "                #self.ins_seqs[new_ins] = cur_cnt + 1\n",
    "                if new_indel in self.read_vars : self.read_vars.update({new_indel:self.read_vars[new_indel]+1})\n",
    "                else : self.read_vars.update({new_indel:1})\n",
    "                self.read_vars.update({'.':self.read_vars['.']-1}) #need to substitute number of matches by one\n",
    "                \n",
    "                skip_index = i + num_len + indel_len +1\n",
    "            \n",
    "            if(read_str[i] == '.'): self.read_vars.update({'.' :self.read_vars['.']+1})\n",
    "            if(read_str[i] == 'A'): self.read_vars.update({'A' :self.read_vars['A']+1})\n",
    "            if(read_str[i] == 'T'): self.read_vars.update({'T' :self.read_vars['T']+1})\n",
    "            if(read_str[i] == 'C'): self.read_vars.update({'C' :self.read_vars['C']+1})\n",
    "            if(read_str[i] == 'G'): self.read_vars.update({'G' :self.read_vars['G']+1})\n",
    "        \n",
    "        self.read_vars = OrderedDict(sorted(self.read_vars.items(), key=lambda item: item[1], reverse=True))\n",
    "        self.read_vars_list = list(self.read_vars.items())\n",
    "        \n",
    "    \n",
    "    \n",
    "    def call_variants_binomial(self, var1, var2):\n",
    "        result_line = ''\n",
    "        genotype_str = ''\n",
    "        \n",
    "        ref_field = 'ERROR'\n",
    "        alt_field = 'ERROR'\n",
    "        gt_field = 'ERROR'\n",
    "        info_field = 'DP=' + var1.DP\n",
    "        \n",
    "        if var2 is None:   #single variant call      \n",
    "#             print('single variant: ' + var1.var_symbol)\n",
    "            ref_field = self.ref_str\n",
    "            alt_field = var1.var_symbol\n",
    "            if(var1.var_symbol == '.'):\n",
    "                gt_field = '0/0:'+var1.VAF\n",
    "            else:\n",
    "                gt_field = '1/1:'+var1.VAF\n",
    "                \n",
    "                \n",
    "                \n",
    "        else:   #binomial calculation of two most probable calls     \n",
    "            var1, var2 = calculate_binomial(var1, var2, p)\n",
    "            \n",
    "            if(var2 is None): #one call is the most probable one\n",
    "#                 print('binomial, single variant: ' + var1.var_symbol)\n",
    "                if(var1.var_type == 'INS' or var1.var_type == 'DEL'):\n",
    "                    info_field = 'INDEL;' + info_field\n",
    "                if(var1.var_type == 'DEL'): \n",
    "                    ref_field = var1.var_symbol\n",
    "                    alt_field = self.ref_str\n",
    "                    gt_field = '1/1:'+var1.VAF\n",
    "                else:\n",
    "                    ref_field = self.ref_str\n",
    "                    alt_field = var1.var_symbol\n",
    "                    if(var1.var_symbol == '.'):\n",
    "                        gt_field = '0/0:'+var1.VAF\n",
    "                    else:\n",
    "                        gt_field = '1/1:'+var1.VAF\n",
    "                        \n",
    "            else: #two calls are most probable\n",
    "#                 print('binomial, double variant: '+var1.var_symbol + '/' + var2.var_symbol)\n",
    "                if(var1.var_type == 'INS' or var1.var_type == 'DEL' or var2.var_type == 'INS' or var2.var_type == 'DEL'):\n",
    "                    info_field = 'INDEL;' + info_field\n",
    "                    \n",
    "                if(var1.var_type == 'DEL' and var2.var_type == 'DEL'):\n",
    "                    ref_field = var1.var_symbol\n",
    "                    alt_field = self.ref_str\n",
    "                    gt_field = '1/1:'+var1.VAF\n",
    "                    result_line += self.seq_str + '\\t' + self.pos_str + '\\t' + '.' + '\\t' + ref_field + '\\t' \n",
    "                    result_line += alt_field + '\\t' + '.' + '\\t' + 'PASS' + '\\t' + info_field + '\\t' + 'GT:VAF' + '\\t' + gt_field +  '\\n'\n",
    "                    \n",
    "                    ref_field = var2.var_symbol\n",
    "                    alt_field = self.ref_str\n",
    "                    gt_field = '1/1:'+var2.VAF\n",
    "                    \n",
    "                elif(var1.var_type == 'DEL' and var2.var_type != 'DEL'):\n",
    "                    ref_field = var1.var_symbol\n",
    "                    alt_field = self.ref_str\n",
    "                    gt_field = '0/0:'+var1.VAF\n",
    "                    result_line += self.seq_str + '\\t' + self.pos_str + '\\t' + '.' + '\\t' + ref_field + '\\t' \n",
    "                    result_line += alt_field + '\\t' + '.' + '\\t' + 'PASS' + '\\t' + info_field + '\\t' + 'GT:VAF' + '\\t' + gt_field +  '\\n'\n",
    "                    \n",
    "                    info_field = 'DP=' + var1.DP\n",
    "                    ref_field = self.ref_str\n",
    "                    alt_field = var2.var_symbol\n",
    "                    if(var2.var_symbol == '.'):\n",
    "                        gt_field = '0/0:' + var2.VAF\n",
    "                    else:\n",
    "                        gt_field = '0/1:' + var2.VAF\n",
    "                \n",
    "                elif(var1.var_type != 'DEL' and var2.var_type == 'DEL'):\n",
    "                    ref_field = var2.var_symbol\n",
    "                    alt_field = self.ref_str\n",
    "                    gt_field = '0/0:'+var2.VAF\n",
    "                    result_line += self.seq_str + '\\t' + self.pos_str + '\\t' + '.' + '\\t' + ref_field + '\\t' \n",
    "                    result_line += alt_field + '\\t' + '.' + '\\t' + 'PASS' + '\\t' + info_field + '\\t' + 'GT:VAF' + '\\t' + gt_field +  '\\n'\n",
    "                    \n",
    "                    info_field = 'DP=' + var1.DP\n",
    "                    ref_field = self.ref_str\n",
    "                    alt_field = var1.var_symbol\n",
    "                    if(var1.var_symbol == '.'):\n",
    "                        gt_field = '0/0:'+var1.VAF\n",
    "                    else:\n",
    "                        gt_field = '0/1:'+var1.VAF\n",
    "                        \n",
    "                else: # (var1.var_type != 'DEL' && var2.var_type != 'DEL')\n",
    "                    ref_field = self.ref_str\n",
    "                    if(var1.var_symbol == '.'):\n",
    "                        alt_field = var2.var_symbol\n",
    "                        gt_field = '0/1:' + var1.VAF + ',' + var2.VAF\n",
    "                    elif (var2.var_symbol == '.'):\n",
    "                        alt_field = var1.var_symbol\n",
    "                        gt_field = '0/1:' + var2.VAF + ',' + var1.VAF\n",
    "                    else:\n",
    "                        alt_field = var1.var_symbol + ',' + var2.var_symbol\n",
    "                        gt_field = '1/2:'+var1.VAF + ',' + var2.VAF\n",
    "                            \n",
    "        assert (ref_field != 'ERROR'), 'Could not calculate ref field!'\n",
    "        assert (alt_field != 'ERROR'), 'Could not calculate alt field!'\n",
    "        assert (gt_field != 'ERROR'), 'Could not calculate gt field!'\n",
    "        \n",
    "        \n",
    "        result_line += self.seq_str + '\\t' + self.pos_str + '\\t' + '.' + '\\t' + ref_field + '\\t' \n",
    "        result_line += alt_field + '\\t' + '.' + '\\t' + 'PASS' + '\\t' + info_field + '\\t' + 'GT:VAF' + '\\t' + gt_field +  '\\n'\n",
    "        \n",
    "        return result_line\n",
    "    \n",
    "\n",
    "#tests\n",
    "p = 0.85\n",
    "test_pileup = Pileup_line('21\t9483407\tG\t54\t..,.,.,.,AAAA.,.,,,,,,..,.,.AA..,,,..T.,,,,,,,,,,,^!,^!,\ta?AbAiEiFgFmgSH;IIIGC@ZnHhI[gFFAIHG>jFFHIIIBBHEEIIG8A')\n",
    "test_res_str = test_pileup.process_read_line()\n",
    "assert (test_pileup.var1.var_symbol == '.' and test_pileup.var1.num_occ == 45), 'Incorrect number of symbols found'\n",
    "assert (test_pileup.var2.var_symbol == 'A' and test_pileup.var2.num_occ == 6), 'Incorrect number of symbols found'\n",
    "res_str = '21\\t9483407\\t.\\tG\\t.\\t.\\tPASS\\tDP=54\\tGT:VAF\\t0/0:0.8\\n'\n",
    "assert (test_res_str == res_str), 'Incorrect output string'\n",
    "\n",
    "test_pileup2 = Pileup_line('21\t9483320\tA\t22\t...+2AT.+2AT..+2AT.+2AT...,.+2AT...+2AT..+2AT..,.^!.\tBFFECEE>E<CCC?BBAB??@>')\n",
    "test_res_str2 = test_pileup2.process_read_line()\n",
    "assert (test_pileup2.var1.var_symbol == '.' and test_pileup2.var1.num_occ == 15), 'Incorrect number of symbols found'\n",
    "assert (test_pileup2.var2.var_symbol == 'AAT' and test_pileup2.var2.num_occ == 7), 'Incorrect number of symbols found'\n",
    "assert (test_pileup2.read_vars['.'] == 15), 'Incorrect number of symbols found'\n",
    "assert (test_pileup2.read_vars['+2AT'] == 7), 'Incorrect number of symbols found'\n",
    "res_str2 = '21\\t9483320\\t.\\tA\\tAAT\\t.\\tPASS\\tINDEL;DP=22\\tGT:VAF\\t0/1:0.7,0.3\\n'\n",
    "assert (test_res_str2 == res_str2), 'Incorrect output string'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "##\n",
    "## MAIN Function - goes through the pileup file, writes to .vcf file\n",
    "## \n",
    "    \n",
    "def main_func(pileup_file_str, output_file_str):\n",
    "    with open (pileup_file_str, 'r') as pileup_file, open (output_file_str, 'w') as temp, open (reference_fai_file_str, 'r') as fai:\n",
    "\n",
    "        temp = write_vcf_header(temp, fai)\n",
    "\n",
    "        line_num = 0;\n",
    "        for line in iter(pileup_file.readline, ''):\n",
    "            line_num+=1 \n",
    "            #if (line_num > 3): break\n",
    "            base_pileup = Pileup_line(line)\n",
    "            if(base_pileup.cnt > 0):\n",
    "                res_str = str(base_pileup.res_str)\n",
    "                processed_str = base_pileup.process_read_line()\n",
    "                temp.write(processed_str)       \n",
    "\n",
    "        print('Processing done.' + output_file_str + ' file created.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing done.variant_call_short_results/binom_p0.5_short_variant_call.vcf file created.\n",
      "Processing done.variant_call_short_results/binom_p0.6_short_variant_call.vcf file created.\n",
      "Processing done.variant_call_short_results/binom_p0.65_short_variant_call.vcf file created.\n",
      "Processing done.variant_call_short_results/binom_p0.7_short_variant_call.vcf file created.\n",
      "Processing done.variant_call_short_results/binom_p0.75_short_variant_call.vcf file created.\n",
      "Processing done.variant_call_short_results/binom_p0.8_short_variant_call.vcf file created.\n",
      "Processing done.variant_call_short_results/binom_p0.85_short_variant_call.vcf file created.\n",
      "Processing done.variant_call_short_results/binom_p0.9_short_variant_call.vcf file created.\n",
      "Processing done.variant_call_short_results/binom_p0.95_short_variant_call.vcf file created.\n",
      "Processing done.variant_call_short_results/binom_p0.98_short_variant_call.vcf file created.\n",
      "Processing done.variant_call_short_results/binom_p0.99_short_variant_call.vcf file created.\n",
      "Processing took 31.351436853408813 seconds\n"
     ]
    }
   ],
   "source": [
    "##\n",
    "## MAIN CODE - opens files, calls main task\n",
    "##\n",
    "start = time.time()\n",
    "\n",
    "# for i in [0.6]: ########### REMOVE\n",
    "for i in [0.5, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95, 0.98, 0.99]:\n",
    "    i_str = str(i)\n",
    "#     output_file_str = 'variant_call_results/binom_p' + i_str + '_variant_call.vcf'\n",
    "    pileup_file_str = 'final_short_pileup' \n",
    "    output_file_str = 'variant_call_short_results/binom_p' + i_str + '_short_variant_call.vcf'\n",
    "    p = i\n",
    "    main_func(pileup_file_str, output_file_str)\n",
    "                \n",
    "end = time.time() \n",
    "print('Processing took ' + str(end - start) + ' seconds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing done.algo_test_pileup_variant_call.vcf file created.\n",
      "Processing took 0.008124828338623047 seconds\n"
     ]
    }
   ],
   "source": [
    "##\n",
    "## TESTS - test pileup file\n",
    "##\n",
    "\n",
    "start = time.time()\n",
    "\n",
    "for i in [0.99]: ########### REMOVE\n",
    "    i_str = str(i)\n",
    "    pileup_file_str = 'algo_test_pileup' \n",
    "    output_file_str = 'algo_test_pileup_variant_call.vcf'\n",
    "    p = i\n",
    "    main_func(pileup_file_str, output_file_str)\n",
    "                \n",
    "end = time.time() \n",
    "print('Processing took ' + str(end - start) + ' seconds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
